### ISSUE TITLE
Can we return dictionaries from pipeline.define_graph()

### ISSUE THREAD

**Issue by shecker** [NVIDIA/DALI#1385]
Hi
I have quite the "complex" data sample structure with one sample having several different length nested dictionaries. 

_sample:_
+-- 'images_front'
|     +-- 'img0': (jpeg)
|     +-- 'img1': (jpeg)
|     +-- 'img2': (jpeg)
+-- 'images_left': (jpeg)
+-- 'images_right': (jpeg)
+-- 'map_data'
|     +-- 'map0': (numpy array)
|     +-- 'map1': (numpy array)
|     +-- 'map2': (numpy array)
|     +-- 'map3': (numpy array)

I was hoping to be able to use DALI with such a nested data structure, but I am unable to return dictionaries from the define_graph() method.
Instead I get a TypeError. 
> TypeError: Expected outputs of type EdgeReference. Received output type dict

Is there any way to use DALI with such a data structure, I would like to leverage the mixed JPEG decoding and augmentation.

Thanks a lot!
Simon

**Comment by shecker** [comment]
Hi
I have a custom ExternalSource iterator that returns `(inputs, labels)`. 
`inputs` is a nested dictionary containing either jpeg images using `np.frombuffer`  (that need to be decoded) and/or standard numpy arrays that do not need further processing. `inputs` has the format that I described above with variable number of images or numpy arrays depending on the experiment (ie. one experiment may only have images_front/img0, whereas the next may have all three (img0, img1, img2)).

I then try to define a pipeline that is able to handle this dictionary and apply the appropriate functions to each type of data. (nvJPEGDecoder, Resize, etc.).
So basically the pipeline is able to handle and process dictionary structured data and return dictionary structured data. 
Let me know if you need more info. I can also add code.
Thanks

**Comment by JanuszL** [comment]
Hi,
If your the amount of images per batch is the same in per experiment you can use the ExternalSource operator that will read raw numpy arrays and the file reader with prebaked file_list with the list of images you want to put through the decoder.
ExternalSource FileReader
       |                    |
       |                ImageDecoder
       |                    |
      \/                   \/
   Output1          Output2

If it can vary iteration to iteration, you can use the DALI pipeline only for decoding and have a separate one for numpy arrays. Then run the one you want the output from. Currently, it is not possible to dynamically bypass operators in the runtime.

**Comment by JanuszL** [comment]
If it still doesn't work please reopen.
